guix: Pin Guix using `guix time-machine`

dongcarl opened the pull request:
An alternative to #16519, pinning our version of Guix and eliminating a `guix pull` and changing the default Guix profile of builders.

I think this method might be superior, as it:
- Eliminates the possibility of future changes to the `guix environment` command line interface breaking our builds
- Eliminates the need to set up a separate channel repo

It is a more general pinning solution than #16519.

-----

The reason why I didn't originally propose this is because `guix time-machine` is a recent addition to Guix, only available since `f675f8dec73d02e319e607559ed2316c299ae8c7`

dongcarl committed:
guix: Pin Guix using `guix time-machine`

MarcoFalke commented:
> Eliminates the need to set up a separate channel repo

Wasn't the conclusion that the channel could be in the same repo? See https://github.com/bitcoin/bitcoin/pull/16519#issuecomment-531843141

Though, I guess the time machine still makes sense when we can switch it to vanilla upstream and not use `dongcarl/guix`?

dongcarl commented:
> Wasn't the conclusion that the channel could be in the same repo? See [#16519 (comment)](https://github.com/bitcoin/bitcoin/pull/16519#issuecomment-531843141)
>
> Though, I guess the time machine still makes sense when we can switch it to vanilla upstream and not use `dongcarl/guix`?

Ah sorry, what I meant to say is that using `time-machine` eliminates the need for any channel/inferior whatsoever! You define some extra packages in the manifest, and just point directly to the version of Guix you want. This is why this diff is so small.

Previously, we had to point to the channel (which contained some extra packages), which in turn pointed to the version of Guix we wanted.

Side benefit: if we used channels in `bitcoin/bitcoin` or in `bitcoin-core/packaging`, we would need a `.guix-channel` file at the root of the tree. This method doesn't.

MarcoFalke commented:
Approach ACK 4763171ff7ad03e654ccd9f78b5ac5ab17aecd25

Didn't test or review the diff to guix upstream

fanquake commented:
Concept ACK

> Side benefit: if we used channels in bitcoin/bitcoin or in bitcoin-core/packaging, we would need a .guix-channel file at the root of the tree. This method doesn't.

Big ++ if this means the project doesn't have to maintain more build / packaging infrastructure.

 I've completed a Guix build using this PR (https://github.com/bitcoin/bitcoin/pull/17933/commits/4763171ff7ad03e654ccd9f78b5ac5ab17aecd25) and this [Docker container](https://github.com/fanquake/core-review/tree/master/guix):

```bash
env PATH="/root/.config/guix/current/bin${PATH:+:}$PATH" guix describe
  guix 65b510b
    repository URL: https://git.savannah.gnu.org/git/guix.git
    branch: master
    commit: 65b510bbc4f2a9ce5bfe3355e6006e9d08f14532

git rev-parse HEAD
4763171ff7ad03e654ccd9f78b5ac5ab17aecd25

find output/ -type f -print0 | sort -z | xargs -r0 sha256sum
edf4db52c0835a89dec1b9c90d7fe43bf94a7ffaa0a9cf54b528bf97283af256  output/bitcoin-0.19.99-aarch64-linux-gnu-debug.tar.gz
816d5a669333ee84a665b8ec8873a288123f4f0008b2943461393f27aacbb7bc  output/bitcoin-0.19.99-aarch64-linux-gnu.tar.gz
3b211b8c7056c81afba457f56f5fbc8a93d819880fac5e464d9a49d82aa87d60  output/bitcoin-0.19.99-arm-linux-gnueabihf-debug.tar.gz
7ae65539e3866af01e843b1008bd604dccf0c327a811a5fbff749a0a0ac3ebbe  output/bitcoin-0.19.99-arm-linux-gnueabihf.tar.gz
0215604cd6706a58a798112bd5284ee2b4d8c941a2da2ba129eefd6f9b615fb1  output/bitcoin-0.19.99-i686-linux-gnu-debug.tar.gz
1a19ffc68a191c5246e8ed30c34548c9ac4351bcf4070b29799144e60cf43476  output/bitcoin-0.19.99-i686-linux-gnu.tar.gz
6632680da7511d941e40a7663246d167b5ce5a58924c9923790988707d71f628  output/bitcoin-0.19.99-riscv64-linux-gnu-debug.tar.gz
b3acdb8db33d77cabe99198e4db2b94757b5bfb5b68b29568ffd432a6d6380d6  output/bitcoin-0.19.99-riscv64-linux-gnu.tar.gz
dab9bf161c283d996b4bf168c7365d966375b434d2a329fbfeb02ee252dba824  output/bitcoin-0.19.99-x86_64-linux-gnu-debug.tar.gz
7ba0242c970eb466a20cfc015f88fb472efba08df5c25b9d668878515a420d07  output/bitcoin-0.19.99-x86_64-linux-gnu.tar.gz
4ac1efb49459a1bb98216580ca3f05c367b8628440505141e924c8de6691639e  output/src/bitcoin-0.19.99.tar.gz
```

MarcoFalke commented:
Why is this a draft?

fanquake reviewed (commented):
> The reason why I didn't originally propose this is because guix time-machine is a recent addition to Guix, only available since f675f8dec73d02e319e607559ed2316c299ae8c7

Given that Guix is a new addition to the project, and there's only a few people doing builds, I've got no issue with us requiring Guix master for now.

I've completed `time-machine` builds for [Linux](https://github.com/bitcoin/bitcoin/pull/17933#issuecomment-575056016) and [Windows](https://github.com/bitcoin/bitcoin/pull/17595#issuecomment-577998443). macOS builds are also working as far as I know; I haven't done a build yet, because I haven't got an SDK with the headers packaged.

I've taken a quick look at the [2019-10-bitcoin-staging-GREAT-CORE-UPDATES-MERGE](https://github.com/dongcarl/guix/tree/2019-10-bitcoin-staging-GREAT-CORE-UPDATES-MERGE) branch as well. I can see the changes you've made match some upstream discussion on the Guix mailing list [here](https://lists.gnu.org/archive/html/guix-patches/2019-11/msg00579.html). 

It also seems that your commit with [David Truby's fix for libstdc++](https://github.com/dongcarl/guix/commit/efdf3dd37413041bb1db74e140be326d6fff1a2f) has now made it into Guix master [496bbeee7a0afbf1c7cbcc0de8c33c2111cb661f](https://git.savannah.gnu.org/cgit/guix.git/commit/?id=496bbeee7a0afbf1c7cbcc0de8c33c2111cb661f). So I guess will be dropped when you pull upstream.

@dongcarl I'd be pretty happy for us to merge this change. Will ACK once you mark as ready for review.

dongcarl commented:
> I've completed a Guix build using this PR ([4763171](https://github.com/bitcoin/bitcoin/commit/4763171ff7ad03e654ccd9f78b5ac5ab17aecd25)) and this [Docker container](https://github.com/fanquake/core-review/tree/master/guix) [](/fanquake/core-review/tree/HEAD@{2020-01-16T09:12:47Z}/guix):
> 
> ```shell
> env PATH="/root/.config/guix/current/bin${PATH:+:}$PATH" guix describe
>   guix 65b510b
>     repository URL: https://git.savannah.gnu.org/git/guix.git
>     branch: master
>     commit: 65b510bbc4f2a9ce5bfe3355e6006e9d08f14532
> 
> git rev-parse HEAD
> 4763171ff7ad03e654ccd9f78b5ac5ab17aecd25
> 
> find output/ -type f -print0 | sort -z | xargs -r0 sha256sum
> edf4db52c0835a89dec1b9c90d7fe43bf94a7ffaa0a9cf54b528bf97283af256  output/bitcoin-0.19.99-aarch64-linux-gnu-debug.tar.gz
> 816d5a669333ee84a665b8ec8873a288123f4f0008b2943461393f27aacbb7bc  output/bitcoin-0.19.99-aarch64-linux-gnu.tar.gz
> 3b211b8c7056c81afba457f56f5fbc8a93d819880fac5e464d9a49d82aa87d60  output/bitcoin-0.19.99-arm-linux-gnueabihf-debug.tar.gz
> 7ae65539e3866af01e843b1008bd604dccf0c327a811a5fbff749a0a0ac3ebbe  output/bitcoin-0.19.99-arm-linux-gnueabihf.tar.gz
> 0215604cd6706a58a798112bd5284ee2b4d8c941a2da2ba129eefd6f9b615fb1  output/bitcoin-0.19.99-i686-linux-gnu-debug.tar.gz
> 1a19ffc68a191c5246e8ed30c34548c9ac4351bcf4070b29799144e60cf43476  output/bitcoin-0.19.99-i686-linux-gnu.tar.gz
> 6632680da7511d941e40a7663246d167b5ce5a58924c9923790988707d71f628  output/bitcoin-0.19.99-riscv64-linux-gnu-debug.tar.gz
> b3acdb8db33d77cabe99198e4db2b94757b5bfb5b68b29568ffd432a6d6380d6  output/bitcoin-0.19.99-riscv64-linux-gnu.tar.gz
> dab9bf161c283d996b4bf168c7365d966375b434d2a329fbfeb02ee252dba824  output/bitcoin-0.19.99-x86_64-linux-gnu-debug.tar.gz
> 7ba0242c970eb466a20cfc015f88fb472efba08df5c25b9d668878515a420d07  output/bitcoin-0.19.99-x86_64-linux-gnu.tar.gz
> 4ac1efb49459a1bb98216580ca3f05c367b8628440505141e924c8de6691639e  output/src/bitcoin-0.19.99.tar.gz
> ```

Fantastic, see my matching hashes below. Note that even though our `guix describe`'s differ, we get the same results! `guix time-machine` FTW!
```shell
$ guix describe
Generation 36   Jan 14 2020 14:50:02    (current)
  guix 09bad3c
    repository URL: https://git.savannah.gnu.org/git/guix.git
    branch: master
    commit: 09bad3cd42ca51974f3697f04ce841bf2fe6786c

$ git rev-parse HEAD
4763171ff7ad03e654ccd9f78b5ac5ab17aecd25

$ find output/ -type f -print0 | sort -z | xargs -r0 sha256sum
edf4db52c0835a89dec1b9c90d7fe43bf94a7ffaa0a9cf54b528bf97283af256  output/bitcoin-0.19.99-aarch64-linux-gnu-debug.tar.gz
816d5a669333ee84a665b8ec8873a288123f4f0008b2943461393f27aacbb7bc  output/bitcoin-0.19.99-aarch64-linux-gnu.tar.gz
3b211b8c7056c81afba457f56f5fbc8a93d819880fac5e464d9a49d82aa87d60  output/bitcoin-0.19.99-arm-linux-gnueabihf-debug.tar.gz
7ae65539e3866af01e843b1008bd604dccf0c327a811a5fbff749a0a0ac3ebbe  output/bitcoin-0.19.99-arm-linux-gnueabihf.tar.gz
0215604cd6706a58a798112bd5284ee2b4d8c941a2da2ba129eefd6f9b615fb1  output/bitcoin-0.19.99-i686-linux-gnu-debug.tar.gz
1a19ffc68a191c5246e8ed30c34548c9ac4351bcf4070b29799144e60cf43476  output/bitcoin-0.19.99-i686-linux-gnu.tar.gz
6632680da7511d941e40a7663246d167b5ce5a58924c9923790988707d71f628  output/bitcoin-0.19.99-riscv64-linux-gnu-debug.tar.gz
b3acdb8db33d77cabe99198e4db2b94757b5bfb5b68b29568ffd432a6d6380d6  output/bitcoin-0.19.99-riscv64-linux-gnu.tar.gz
dab9bf161c283d996b4bf168c7365d966375b434d2a329fbfeb02ee252dba824  output/bitcoin-0.19.99-x86_64-linux-gnu-debug.tar.gz
7ba0242c970eb466a20cfc015f88fb472efba08df5c25b9d668878515a420d07  output/bitcoin-0.19.99-x86_64-linux-gnu.tar.gz
4ac1efb49459a1bb98216580ca3f05c367b8628440505141e924c8de6691639e  output/src/bitcoin-0.19.99.tar.gz
```

dongcarl committed:
guix: Update documentation for time-machine

Wait a minute, doc. Are you telling me you built a time machine... Out
of a functional package manager?

dongcarl commented:
Marking as "Ready for review": 88c83636d5a56bd9551577139786bdd3e74852c2

```
$ find output/ -type f -print0 | sort -z | xargs -r0 sha256sum
39608e90c90be13c10488c25fb4bfa3969c5304a4fcde21ce8c57acc0947c9d8  output/bitcoin-0.19.99-aarch64-linux-gnu-debug.tar.gz
0540ab5aba8959c45255b60077ee7498333852f3d2067610357d811925762f61  output/bitcoin-0.19.99-aarch64-linux-gnu.tar.gz
cc1fcc9dd71162baa0f8d85ae9b514bdbf7f92001bfbd45998dd6db92260edc0  output/bitcoin-0.19.99-arm-linux-gnueabihf-debug.tar.gz
f978a4496dd37aa7588e7868ae52607b7e8849d86d27b82f2007ea369cddd1fc  output/bitcoin-0.19.99-arm-linux-gnueabihf.tar.gz
92da6f327f187cc44bd30c31852111bf09879640efac630303bad2f040e20b93  output/bitcoin-0.19.99-i686-linux-gnu-debug.tar.gz
db4654620e0dfe03c4f3a3570b1a9cace15c28c9f68497614710156b7a4a1318  output/bitcoin-0.19.99-i686-linux-gnu.tar.gz
d841618b51ce5e9a83ca789b89e86b617c646fee2c9d15487e521e29741f4e72  output/bitcoin-0.19.99-riscv64-linux-gnu-debug.tar.gz
879e526aa5139eea666c949036d811652c3c6f7e3601a53a7617bc21738baf1a  output/bitcoin-0.19.99-riscv64-linux-gnu.tar.gz
5cde323bbcf1c3331be294b562aab477a72adfad141d5e019ff10e1753178de8  output/bitcoin-0.19.99-x86_64-linux-gnu-debug.tar.gz
8cf0240185043fdcfb41500058bbc60c7ec61d15ec30230cc9a33c8e99d73955  output/bitcoin-0.19.99-x86_64-linux-gnu.tar.gz
843f958009812087fcb68fab8117d7710db19bcc64d36966121975718609dced  output/src/bitcoin-0.19.99.tar.gz
```

#### Changes since fanquake's last build

1. Updated my Guix branch to have clearer commit messages.
2. Added commit adjusting the `contrib/guix/README.md` according to this PR.


#### Some context on my Guix branch

Outstanding commits between my Guix branch and Guix master: https://github.com/dongcarl/guix/compare/377606238f0260a849a1f529d94e1546276a5db6...b3a7c72c8b2425f8ddb0fc6e3b1caeed40f86dee

Proof that `3776062` is in Guix master: https://git.savannah.gnu.org/cgit/guix.git/commit/?id=377606238f0260a849a1f529d94e1546276a5db6

fanquake reviewed (approved):
ACK 88c83636d5a56bd9551577139786bdd3e74852c2

Ran another build and seeing (new) matching results:

```bash
bash-5.0# env PATH="/root/.config/guix/current/bin${PATH:+:}$PATH" guix describe
  guix 7ee8acb
    repository URL: https://git.savannah.gnu.org/git/guix.git
    branch: master
    commit: 7ee8acbb76bd440a8985a4b2a20b75521c6853ed

bash-5.0# git rev-parse HEAD
88c83636d5a56bd9551577139786bdd3e74852c2

bash-5.0# find output/ -type f -print0 | sort -z | xargs -r0 sha256sum
39608e90c90be13c10488c25fb4bfa3969c5304a4fcde21ce8c57acc0947c9d8  output/bitcoin-0.19.99-aarch64-linux-gnu-debug.tar.gz
0540ab5aba8959c45255b60077ee7498333852f3d2067610357d811925762f61  output/bitcoin-0.19.99-aarch64-linux-gnu.tar.gz
cc1fcc9dd71162baa0f8d85ae9b514bdbf7f92001bfbd45998dd6db92260edc0  output/bitcoin-0.19.99-arm-linux-gnueabihf-debug.tar.gz
f978a4496dd37aa7588e7868ae52607b7e8849d86d27b82f2007ea369cddd1fc  output/bitcoin-0.19.99-arm-linux-gnueabihf.tar.gz
92da6f327f187cc44bd30c31852111bf09879640efac630303bad2f040e20b93  output/bitcoin-0.19.99-i686-linux-gnu-debug.tar.gz
db4654620e0dfe03c4f3a3570b1a9cace15c28c9f68497614710156b7a4a1318  output/bitcoin-0.19.99-i686-linux-gnu.tar.gz
d841618b51ce5e9a83ca789b89e86b617c646fee2c9d15487e521e29741f4e72  output/bitcoin-0.19.99-riscv64-linux-gnu-debug.tar.gz
879e526aa5139eea666c949036d811652c3c6f7e3601a53a7617bc21738baf1a  output/bitcoin-0.19.99-riscv64-linux-gnu.tar.gz
5cde323bbcf1c3331be294b562aab477a72adfad141d5e019ff10e1753178de8  output/bitcoin-0.19.99-x86_64-linux-gnu-debug.tar.gz
8cf0240185043fdcfb41500058bbc60c7ec61d15ec30230cc9a33c8e99d73955  output/bitcoin-0.19.99-x86_64-linux-gnu.tar.gz
843f958009812087fcb68fab8117d7710db19bcc64d36966121975718609dced  output/src/bitcoin-0.19.99.tar.gz
```

Had another look over your branch guix branch as well. Do you have any sort of timeline in regards to your patches being upstreamed? We should probably update #16496.

I've also [updated](https://github.com/fanquake/core-review/commit/ade01dc950267749f54cf9794e67b19e028bf9bf) the Guix instructions in [core-review](https://github.com/fanquake/core-review/tree/master/guix) to pull from master rather than your repo.

MarcoFalke commented on contrib/guix/README.md:
```diff
@@ -224,6 +222,7 @@ repository and will likely put one up soon.
 [guix/substitute-server-auth]: https://www.gnu.org/software/guix/manual/en/html_node/Substitute-Server-Authorization.html
 [guix/inferiors]: https://www.gnu.org/software/guix/manual/en/html_node/Inferiors.html
```
Can remove unused link now?

fanquake replied:
Made a note to fix this in 17595.
